ICE: "associated const projection is not supported yet"

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code
`rustc file.rs -Ztrait-solver=next-coherence`
```Rust
pub struct NoPin;

impl Pins<TA> for NoPin {}

pub trait PinA<PER> {
    const A: u8;
}

impl<PER> PinA<PER> for NoPin {
    const A: u8 = 0;
}

pub trait Pins<USART> {}

impl<USART, T> Pins<USART> for T where T: PinA<USART, A = { TA }> {}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.74.0-nightly (9f5fc1bd4 2023-09-02)
binary: rustc
commit-hash: 9f5fc1bd443f59583e7af0d94d289f95fe1e20c4
commit-date: 2023-09-02
host: x86_64-unknown-linux-gnu
release: 1.74.0-nightly
LLVM version: 17.0.0
```

### Error output

```
error[E0412]: cannot find type `TA` in this scope
 --> 8994375BB6FC4D929D8698F40C9320D437205E1BC19A28C63FCECCE9ED592074.rs:3:11
  |
3 | impl Pins<TA> for NoPin {}
  |           ^^ not found in this scope
  |
help: you might be missing a type parameter
  |
3 | impl<TA> Pins<TA> for NoPin {}
  |     ++++

error[E0425]: cannot find value `TA` in this scope
  --> 8994375BB6FC4D929D8698F40C9320D437205E1BC19A28C63FCECCE9ED592074.rs:15:61
   |
15 | impl<USART, T> Pins<USART> for T where T: PinA<USART, A = { TA }> {}
   |                                                             ^^ not found in this scope

error[E0658]: associated const equality is incomplete
  --> 8994375BB6FC4D929D8698F40C9320D437205E1BC19A28C63FCECCE9ED592074.rs:15:55
   |
15 | impl<USART, T> Pins<USART> for T where T: PinA<USART, A = { TA }> {}
   |                                                       ^^^^^^^^^^
   |
   = note: see issue #92827 <https://github.com/rust-lang/rust/issues/92827> for more information
   = help: add `#![feature(associated_const_equality)]` to the crate attributes to enable

error[E0601]: `main` function not found in crate `8994375BB6FC4D929D8698F40C9320D437205E1BC19A28C63FCECCE9ED592074`
  --> 8994375BB6FC4D929D8698F40C9320D437205E1BC19A28C63FCECCE9ED592074.rs:15:69
   |
15 | impl<USART, T> Pins<USART> for T where T: PinA<USART, A = { TA }> {}
   |                                                                     ^ consider adding a `main` function to `8994375BB6FC4D929D8698F40C9320D437205E1BC19A28C63FCECCE9ED592074.rs`
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_trait_selection/src/solve/project_goals.rs:247:41: associated const projection is not supported yet

thread 'rustc' panicked at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/compiler/rustc_errors/src/lib.rs:1651:9:
Box<dyn Any>
stack backtrace:
   0:     0x7f278b0f3dec - std::backtrace_rs::backtrace::libunwind::trace::hdcb9c0b22500da91
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f278b0f3dec - std::backtrace_rs::backtrace::trace_unsynchronized::h91b351fd34050178
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f278b0f3dec - std::sys_common::backtrace::_print_fmt::he469089ab3fbd8a5
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7f278b0f3dec - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h57d5e31c47a8a531
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f278b15a05c - core::fmt::rt::Argument::fmt::h6e8bafe434c307b3
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/core/src/fmt/rt.rs:138:9
   5:     0x7f278b15a05c - core::fmt::write::hbf1ef8babec800ef
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/core/src/fmt/mod.rs:1094:21
   6:     0x7f278b0e68fe - std::io::Write::write_fmt::h3273b2319fbd2ddc
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/io/mod.rs:1714:15
   7:     0x7f278b0f3bd4 - std::sys_common::backtrace::_print::h9f429c8b9622817f
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f278b0f3bd4 - std::sys_common::backtrace::print::h3e62e3dc9cc0960c
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f278b0f6cea - std::panicking::panic_hook_with_disk_dump::{{closure}}::h336ab5ed5f752f5b
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/panicking.rs:280:22
  10:     0x7f278b0f69e5 - std::panicking::panic_hook_with_disk_dump::h0b2ed72036e73f37
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/panicking.rs:314:9
  11:     0x7f2789f4d079 - rustc_driver_impl[bd3214347f952018]::install_ice_hook::{closure#0}
  12:     0x7f278b0f75a3 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hf5daf4aeb8dae1f0
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/alloc/src/boxed.rs:2021:9
  13:     0x7f278b0f75a3 - std::panicking::rust_panic_with_hook::h725e440d3f428ec3
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/panicking.rs:757:13
  14:     0x7f278a4a7d94 - std[15568f5d89cfcf8b]::panicking::begin_panic::<rustc_errors[c5bbdc7440f4c02c]::ExplicitBug>::{closure#0}
  15:     0x7f278a4a5076 - std[15568f5d89cfcf8b]::sys_common::backtrace::__rust_end_short_backtrace::<std[15568f5d89cfcf8b]::panicking::begin_panic<rustc_errors[c5bbdc7440f4c02c]::ExplicitBug>::{closure#0}, !>
  16:     0x7f278a509456 - std[15568f5d89cfcf8b]::panicking::begin_panic::<rustc_errors[c5bbdc7440f4c02c]::ExplicitBug>
  17:     0x7f278a463994 - <rustc_errors[c5bbdc7440f4c02c]::HandlerInner>::bug::<alloc[8c08642181f67925]::string::String>
  18:     0x7f278a463826 - <rustc_errors[c5bbdc7440f4c02c]::Handler>::bug::<alloc[8c08642181f67925]::string::String>
  19:     0x7f278a4d738c - rustc_middle[a400c0366bd3264d]::util::bug::opt_span_bug_fmt::<rustc_span[d6219d1cb1a2cd02]::span_encoding::Span>::{closure#0}
  20:     0x7f278a4d616a - rustc_middle[a400c0366bd3264d]::ty::context::tls::with_opt::<rustc_middle[a400c0366bd3264d]::util::bug::opt_span_bug_fmt<rustc_span[d6219d1cb1a2cd02]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f278a4d6138 - rustc_middle[a400c0366bd3264d]::ty::context::tls::with_context_opt::<rustc_middle[a400c0366bd3264d]::ty::context::tls::with_opt<rustc_middle[a400c0366bd3264d]::util::bug::opt_span_bug_fmt<rustc_span[d6219d1cb1a2cd02]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f27881e4920 - rustc_middle[a400c0366bd3264d]::util::bug::bug_fmt
  23:     0x7f278ab98007 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::probe::ProbeCtxt<<rustc_middle[a400c0366bd3264d]::ty::ProjectionPredicate as rustc_trait_selection[f50372471fdcdd04]::solve::assembly::GoalKind>::consider_impl_candidate::{closure#0}, core[afd64c1140d34ddb]::result::Result<rustc_middle[a400c0366bd3264d]::infer::canonical::Canonical<rustc_middle[a400c0366bd3264d]::traits::solve::Response>, rustc_middle[a400c0366bd3264d]::traits::query::NoSolution>>>::enter::<<rustc_middle[a400c0366bd3264d]::ty::ProjectionPredicate as rustc_trait_selection[f50372471fdcdd04]::solve::assembly::GoalKind>::consider_impl_candidate::{closure#1}>
  24:     0x7f278aae583f - <rustc_middle[a400c0366bd3264d]::ty::ProjectionPredicate as rustc_trait_selection[f50372471fdcdd04]::solve::assembly::GoalKind>::consider_impl_candidate
  25:     0x7f278abd1c29 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::assemble_non_blanket_impl_candidates::<rustc_middle[a400c0366bd3264d]::ty::ProjectionPredicate>::{closure#0}
  26:     0x7f278abc98c5 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::assemble_candidates_via_self_ty::<rustc_middle[a400c0366bd3264d]::ty::ProjectionPredicate>
  27:     0x7f278abce161 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::enter_canonical::<core[afd64c1140d34ddb]::result::Result<rustc_middle[a400c0366bd3264d]::infer::canonical::Canonical<rustc_middle[a400c0366bd3264d]::traits::solve::Response>, rustc_middle[a400c0366bd3264d]::traits::query::NoSolution>, <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}::{closure#0}>
  28:     0x7f278aba47fe - <rustc_trait_selection[f50372471fdcdd04]::solve::search_graph::SearchGraph>::with_new_goal::<<rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>::{closure#2}
  29:     0x7f278aba42dd - <rustc_trait_selection[f50372471fdcdd04]::solve::search_graph::SearchGraph>::with_new_goal::<<rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>
  30:     0x7f278abcf329 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_goal
  31:     0x7f278abd31c2 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::try_evaluate_added_goals
  32:     0x7f278abd2801 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_added_goals_and_make_canonical_response
  33:     0x7f278abcda46 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::enter_canonical::<core[afd64c1140d34ddb]::result::Result<rustc_middle[a400c0366bd3264d]::infer::canonical::Canonical<rustc_middle[a400c0366bd3264d]::traits::solve::Response>, rustc_middle[a400c0366bd3264d]::traits::query::NoSolution>, <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}::{closure#0}>
  34:     0x7f278aba47fe - <rustc_trait_selection[f50372471fdcdd04]::solve::search_graph::SearchGraph>::with_new_goal::<<rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>::{closure#2}
  35:     0x7f278aba42dd - <rustc_trait_selection[f50372471fdcdd04]::solve::search_graph::SearchGraph>::with_new_goal::<<rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>
  36:     0x7f278abcf329 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::evaluate_goal
  37:     0x7f278abcc349 - <rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::EvalCtxt>::enter_root::<core[afd64c1140d34ddb]::result::Result<(bool, rustc_middle[a400c0366bd3264d]::traits::solve::Certainty, alloc[8c08642181f67925]::vec::Vec<rustc_middle[a400c0366bd3264d]::traits::solve::Goal<rustc_middle[a400c0366bd3264d]::ty::Predicate>>), rustc_middle[a400c0366bd3264d]::traits::query::NoSolution>, <rustc_infer[5c7a2123edb7d5c]::infer::InferCtxt as rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::InferCtxtEvalExt>::evaluate_root_goal::{closure#0}::{closure#0}>
  38:     0x7f278ab9b1b4 - <rustc_infer[5c7a2123edb7d5c]::infer::InferCtxt as rustc_trait_selection[f50372471fdcdd04]::solve::eval_ctxt::InferCtxtEvalExt>::evaluate_root_goal
  39:     0x7f278abcfd71 - <rustc_trait_selection[f50372471fdcdd04]::solve::fulfill::FulfillmentCtxt as rustc_infer[5c7a2123edb7d5c]::traits::engine::TraitEngine>::select_where_possible
  40:     0x7f278820121e - <rustc_infer[5c7a2123edb7d5c]::infer::InferCtxt as rustc_trait_selection[f50372471fdcdd04]::traits::query::evaluate_obligation::InferCtxtExt>::evaluate_obligation
  41:     0x7f2789927ea0 - <core[afd64c1140d34ddb]::iter::adapters::map::Map<core[afd64c1140d34ddb]::iter::adapters::flatten::Flatten<core[afd64c1140d34ddb]::array::iter::IntoIter<&alloc[8c08642181f67925]::vec::Vec<rustc_middle[a400c0366bd3264d]::ty::Predicate>, 2usize>>, rustc_trait_selection[f50372471fdcdd04]::traits::coherence::impl_intersection_has_impossible_obligation::{closure#0}> as core[afd64c1140d34ddb]::iter::traits::iterator::Iterator>::try_fold::<(), core[afd64c1140d34ddb]::iter::traits::iterator::Iterator::find::check<rustc_infer[5c7a2123edb7d5c]::traits::Obligation<rustc_middle[a400c0366bd3264d]::ty::Predicate>, &mut rustc_trait_selection[f50372471fdcdd04]::traits::coherence::impl_intersection_has_impossible_obligation::{closure#1}>::{closure#0}, core[afd64c1140d34ddb]::ops::control_flow::ControlFlow<rustc_infer[5c7a2123edb7d5c]::traits::Obligation<rustc_middle[a400c0366bd3264d]::ty::Predicate>>>
  42:     0x7f2789162b2c - rustc_trait_selection[f50372471fdcdd04]::traits::coherence::impl_intersection_has_impossible_obligation
  43:     0x7f2789162a4f - <rustc_trait_selection[f50372471fdcdd04]::traits::select::SelectionContext>::with_treat_inductive_cycle_as::<core[afd64c1140d34ddb]::option::Option<rustc_infer[5c7a2123edb7d5c]::traits::Obligation<rustc_middle[a400c0366bd3264d]::ty::Predicate>>, rustc_trait_selection[f50372471fdcdd04]::traits::coherence::overlap::{closure#0}>
  44:     0x7f2789161b60 - rustc_trait_selection[f50372471fdcdd04]::traits::coherence::overlap
  45:     0x7f278922544d - <rustc_middle[a400c0366bd3264d]::traits::specialization_graph::Children as rustc_trait_selection[f50372471fdcdd04]::traits::specialize::specialization_graph::ChildrenExt>::insert
  46:     0x7f2789223897 - <rustc_middle[a400c0366bd3264d]::traits::specialization_graph::Graph as rustc_trait_selection[f50372471fdcdd04]::traits::specialize::specialization_graph::GraphExt>::insert
  47:     0x7f2789222591 - rustc_trait_selection[f50372471fdcdd04]::traits::specialize::specialization_graph_provider
  48:     0x7f27892e96db - rustc_query_impl[2b96a7f183453ea1]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2b96a7f183453ea1]::query_impl::specialization_graph_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[a400c0366bd3264d]::query::erase::Erased<[u8; 8usize]>>
  49:     0x7f27892e96b1 - <rustc_query_impl[2b96a7f183453ea1]::query_impl::specialization_graph_of::dynamic_query::{closure#2} as core[afd64c1140d34ddb]::ops::function::FnOnce<(rustc_middle[a400c0366bd3264d]::ty::context::TyCtxt, rustc_span[d6219d1cb1a2cd02]::def_id::DefId)>>::call_once
  50:     0x7f2788053d1a - rustc_query_system[f349c9744a75bab0]::query::plumbing::try_execute_query::<rustc_query_impl[2b96a7f183453ea1]::DynamicConfig<rustc_query_system[f349c9744a75bab0]::query::caches::DefaultCache<rustc_span[d6219d1cb1a2cd02]::def_id::DefId, rustc_middle[a400c0366bd3264d]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[2b96a7f183453ea1]::plumbing::QueryCtxt, false>
  51:     0x7f27898eed42 - rustc_query_impl[2b96a7f183453ea1]::query_impl::specialization_graph_of::get_query_non_incr::__rust_end_short_backtrace
  52:     0x7f278957a3e8 - rustc_hir_analysis[f4190e851159bd14]::coherence::coherent_trait
  53:     0x7f2788ba18d1 - rustc_query_impl[2b96a7f183453ea1]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2b96a7f183453ea1]::query_impl::coherent_trait::dynamic_query::{closure#2}::{closure#0}, rustc_middle[a400c0366bd3264d]::query::erase::Erased<[u8; 0usize]>>
  54:     0x7f2788ba18b1 - <rustc_query_impl[2b96a7f183453ea1]::query_impl::coherent_trait::dynamic_query::{closure#2} as core[afd64c1140d34ddb]::ops::function::FnOnce<(rustc_middle[a400c0366bd3264d]::ty::context::TyCtxt, rustc_span[d6219d1cb1a2cd02]::def_id::DefId)>>::call_once
  55:     0x7f2789418245 - rustc_query_system[f349c9744a75bab0]::query::plumbing::try_execute_query::<rustc_query_impl[2b96a7f183453ea1]::DynamicConfig<rustc_query_system[f349c9744a75bab0]::query::caches::DefaultCache<rustc_span[d6219d1cb1a2cd02]::def_id::DefId, rustc_middle[a400c0366bd3264d]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[2b96a7f183453ea1]::plumbing::QueryCtxt, false>
  56:     0x7f2789417f5e - rustc_query_impl[2b96a7f183453ea1]::query_impl::coherent_trait::get_query_non_incr::__rust_end_short_backtrace
  57:     0x7f27891bf488 - <rustc_session[bc10157bc0a34fd]::session::Session>::track_errors::<rustc_hir_analysis[f4190e851159bd14]::check_crate::{closure#3}, ()>
  58:     0x7f27891bd247 - rustc_hir_analysis[f4190e851159bd14]::check_crate
  59:     0x7f27891bc172 - rustc_interface[32fae85809e14be6]::passes::analysis
  60:     0x7f27893c51ca - rustc_query_impl[2b96a7f183453ea1]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2b96a7f183453ea1]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[a400c0366bd3264d]::query::erase::Erased<[u8; 1usize]>>
  61:     0x7f27893c51b9 - <rustc_query_impl[2b96a7f183453ea1]::query_impl::analysis::dynamic_query::{closure#2} as core[afd64c1140d34ddb]::ops::function::FnOnce<(rustc_middle[a400c0366bd3264d]::ty::context::TyCtxt, ())>>::call_once
  62:     0x7f278967e394 - rustc_query_system[f349c9744a75bab0]::query::plumbing::try_execute_query::<rustc_query_impl[2b96a7f183453ea1]::DynamicConfig<rustc_query_system[f349c9744a75bab0]::query::caches::SingleCache<rustc_middle[a400c0366bd3264d]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[2b96a7f183453ea1]::plumbing::QueryCtxt, false>
  63:     0x7f278967e109 - rustc_query_impl[2b96a7f183453ea1]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  64:     0x7f27895793f3 - <rustc_middle[a400c0366bd3264d]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[bd3214347f952018]::run_compiler::{closure#1}::{closure#2}::{closure#6}, core[afd64c1140d34ddb]::result::Result<(), rustc_span[d6219d1cb1a2cd02]::ErrorGuaranteed>>
  65:     0x7f278947d4ec - <rustc_interface[32fae85809e14be6]::interface::Compiler>::enter::<rustc_driver_impl[bd3214347f952018]::run_compiler::{closure#1}::{closure#2}, core[afd64c1140d34ddb]::result::Result<core[afd64c1140d34ddb]::option::Option<rustc_interface[32fae85809e14be6]::queries::Linker>, rustc_span[d6219d1cb1a2cd02]::ErrorGuaranteed>>
  66:     0x7f2789474ce8 - std[15568f5d89cfcf8b]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[32fae85809e14be6]::util::run_in_thread_with_globals<rustc_interface[32fae85809e14be6]::interface::run_compiler<core[afd64c1140d34ddb]::result::Result<(), rustc_span[d6219d1cb1a2cd02]::ErrorGuaranteed>, rustc_driver_impl[bd3214347f952018]::run_compiler::{closure#1}>::{closure#0}, core[afd64c1140d34ddb]::result::Result<(), rustc_span[d6219d1cb1a2cd02]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[afd64c1140d34ddb]::result::Result<(), rustc_span[d6219d1cb1a2cd02]::ErrorGuaranteed>>
  67:     0x7f278947445e - <<std[15568f5d89cfcf8b]::thread::Builder>::spawn_unchecked_<rustc_interface[32fae85809e14be6]::util::run_in_thread_with_globals<rustc_interface[32fae85809e14be6]::interface::run_compiler<core[afd64c1140d34ddb]::result::Result<(), rustc_span[d6219d1cb1a2cd02]::ErrorGuaranteed>, rustc_driver_impl[bd3214347f952018]::run_compiler::{closure#1}>::{closure#0}, core[afd64c1140d34ddb]::result::Result<(), rustc_span[d6219d1cb1a2cd02]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[afd64c1140d34ddb]::result::Result<(), rustc_span[d6219d1cb1a2cd02]::ErrorGuaranteed>>::{closure#1} as core[afd64c1140d34ddb]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  68:     0x7f278b102015 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::he46350f6bfcc0e8a
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/alloc/src/boxed.rs:2007:9
  69:     0x7f278b102015 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h4038c4c7fc344b18
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/alloc/src/boxed.rs:2007:9
  70:     0x7f278b102015 - std::sys::unix::thread::Thread::new::thread_start::h842f0d826bbbe958
                               at /rustc/9f5fc1bd443f59583e7af0d94d289f95fe1e20c4/library/std/src/sys/unix/thread.rs:108:17
  71:     0x7f2786a8c9eb - <unknown>
  72:     0x7f2786b10dfc - <unknown>
  73:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please attach the file at `/tmp/coherence/icemaker_reduced/rustc-ice-2023-09-03T09:29:00.466305121Z-779691.txt` to your bug report

note: compiler flags: -Z trait-solver=next-coherence

query stack during panic:
#0 [specialization_graph_of] building specialization graph of trait `Pins`
#1 [coherent_trait] coherence checking all impls of trait `Pins`
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0412, E0425, E0601, E0658.
For more information about an error, try `rustc --explain E0412`.
```

</p>
</details>

--- a/compiler/rustc_trait_selection/src/solve/project_goals.rs
+++ b/compiler/rustc_trait_selection/src/solve/project_goals.rs
@@ -244,7 +244,21 @@ impl<'tcx> assembly::GoalKind<'tcx> for ProjectionPredicate<'tcx> {
             // Finally we construct the actual value of the associated type.
             let term = match assoc_def.item.kind {
                 ty::AssocKind::Type => tcx.type_of(assoc_def.item.def_id).map_bound(|ty| ty.into()),
-                ty::AssocKind::Const => bug!("associated const projection is not supported yet"),
+                ty::AssocKind::Const => {
+                    if tcx.features().associated_const_equality {
+                        bug!("associated const projection is not supported yet")
+                    } else {
+                        ty::EarlyBinder::bind(
+                            ty::Const::new_error_with_message(
+                                tcx,
+                                tcx.type_of(assoc_def.item.def_id).instantiate_identity(),
+                                DUMMY_SP,
+                                "associated const projection is not supported yet",
+                            )
+                            .into(),
+                        )
+                    }
+                }
                 ty::AssocKind::Fn => unreachable!("we should never project to a fn"),
             };
 

--- a/tests/ui/traits/new-solver/dont-ice-on-assoc-projection.rs
+++ b/tests/ui/traits/new-solver/dont-ice-on-assoc-projection.rs
@@ -1,0 +1,19 @@
+// compile-flags: -Ztrait-solver=next-coherence
+
+// Makes sure we don't ICE on associated const projection when the feature gate
+// is not enabled, since we should avoid encountering ICEs on stable if possible.
+
+trait Bar {
+    const ASSOC: usize;
+}
+impl Bar for () {
+    const ASSOC: usize = 1;
+}
+
+trait Foo {}
+impl Foo for () {}
+impl<T> Foo for T where T: Bar<ASSOC = 0> {}
+//~^ ERROR associated const equality is incomplete
+//~| ERROR conflicting implementations of trait `Foo` for type `()`
+
+fn main() {}

--- a/tests/ui/traits/new-solver/dont-ice-on-assoc-projection.stderr
+++ b/tests/ui/traits/new-solver/dont-ice-on-assoc-projection.stderr
@@ -1,0 +1,21 @@
+error[E0658]: associated const equality is incomplete
+  --> $DIR/dont-ice-on-assoc-projection.rs:15:32
+   |
+LL | impl<T> Foo for T where T: Bar<ASSOC = 0> {}
+   |                                ^^^^^^^^^
+   |
+   = note: see issue #92827 <https://github.com/rust-lang/rust/issues/92827> for more information
+   = help: add `#![feature(associated_const_equality)]` to the crate attributes to enable
+
+error[E0119]: conflicting implementations of trait `Foo` for type `()`
+  --> $DIR/dont-ice-on-assoc-projection.rs:15:1
+   |
+LL | impl Foo for () {}
+   | --------------- first implementation here
+LL | impl<T> Foo for T where T: Bar<ASSOC = 0> {}
+   | ^^^^^^^^^^^^^^^^^ conflicting implementation for `()`
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0119, E0658.
+For more information about an error, try `rustc --explain E0119`.

